Enforce that compilers are single-use.

adolfo-trocoli opened the pull request:
Issue #18 (and my first pull request).

ascopes commented on java-compiler-testing/src/main/java/io/github/ascopes/jct/compilers/AlreadyUsedCompilerException.java:
```diff
@@ -0,0 +1,34 @@
+package io.github.ascopes.jct.compilers;
```
So I think we'll need the license header at the top of each of these files. You can copy-paste that from any other Java file, but Checkstyle will fail without it being here! 

```suggestion
/*
 * Copyright (C) 2022 Ashley Scopes
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.github.ascopes.jct.compilers;
```

ascopes commented on java-compiler-testing/src/main/java/io/github/ascopes/jct/compilers/SimpleCompiler.java:
```diff
@@ -169,6 +170,9 @@ public SimpleCompilation compile() {
     // internally, such as the ECJ global lock that we apply to prevent bugs. Without this, we'd
     // have to mock dozens of additional moving parts. It is difficult to stub super methods
     // if we go down that route.
+	if (compiled)
+		throw new AlreadyUsedCompilerException("There has been a second call to compile() in this Compiler");
```
```suggestion
    if (compiled) {
      throw new AlreadyUsedCompilerException("There has been a second call to compile() in this Compiler");
    }
```

Just for consistency (also uses 2-space indents rather than tabs, which is something I think checkstyle will flag).

ascopes commented on java-compiler-testing/src/test/java/io/github/ascopes/jct/testing/unit/compilers/SimpleCompilerTest.java:
```diff
@@ -74,6 +75,26 @@
+	    // Given
+	    var jsr199Compiler = stub(JavaCompiler.class);
+	    var flagBuilder = stub(FlagBuilder.class);
+	   var compiler = new StubbedCompiler("foobar", jsr199Compiler, flagBuilder);
```
Think checkstyle will raise the indent here. If you use IntelliJ you should be able to just import the project and run the "reformat file" command, it should pick up everything from the `.editorconfig` at the root of the repository.

ascopes reviewed (changes requested):
Looks good! Thanks for picking this up.

Just a few formatting issues that need double checking. Specifically:

- `{ }` around conditions for consistency
- 2 space indents rather than tabs/4 space
- License headers on each file

If possible, could you rebase this onto main (I had to update the checkstyle CI stuff yesterday as I made a silly mistake in something). `git fetch https://github.com/ascopes/jct main && git rebase FETCH_HEAD && git push -f` should be all you need to do in order to do that!

Once that is done, you should be good to go!

adolfo-trocoli commented:
I'm sorry for the delay. I've been having some final exams this weeks. I've commited the suggested changes, but I am not able to rebase the code (or at least I don't know how to do it) because I've modified your project to be able to open it as an Eclipse Project. Because of this, I can not commit every change I've made to the project, so I edited the files manually in the Github Editor. I suppose this isn't the best way to do things but as I said this is my first time doing this. Thank you very much for your patience.

ascopes commented:
@adolfo-trocoli no rush! Exams are always more important. I appreciate the time you have spent on this so far too, it is a great help.

If there are any files you don't want to track from Eclipse, you can add their patterns to the `.gitignore` file before you `git add`.

Try this:

```console
git stash
git fetch https://github.com/ascopes/jct main
git rebase FETCH_HEAD
git stash pop
git push --force-with-lease
```

adolfo-trocoli commented:
Well, I've been trying to do this but once again my knowledge about git comes short.

**Up to the "git stash pop" everything is fine. However, when I input this line, i get:**
_Removing java-compiler-testing/src/main/java/module-info.java
On branch master
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
        new file:   .project
        new file:   examples/.project
        new file:   examples/example-lombok/.project
        new file:   examples/example-serviceloader-with-jpms/.project
        new file:   examples/example-serviceloader/.project
        new file:   java-compiler-testing/.project
        new file:   java-compiler-testing/src/main/java/io/github/ascopes/jct/compilers/AlreadyUsedCompilerException.java
        new file:   java-compiler-testing/src/test/java/io/github/ascopes/jct/testing/unit/compilers/AlreadyUsedCompilerExceptionTest.java

Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   .gitignore
        modified:   java-compiler-testing/src/main/java/io/github/ascopes/jct/compilers/SimpleCompiler.java
        deleted:    java-compiler-testing/src/main/java/module-info.java
        modified:   java-compiler-testing/src/test/java/io/github/ascopes/jct/testing/unit/compilers/SimpleCompilerTest.java

Dropped refs/stash@{0} (56bd3a9e0e0e12c0edfa5532c8792eab522afe8e)_


**With the last line, I get:** 
_fatal: The current branch master has no upstream branch.
To push the current branch and set the remote as upstream, use

    git push --set-upstream origin master_

**But if I try to do this then I get:**
_remote: Permission to ascopes/jct.git denied to adolfo-trocoli.
fatal: unable to access 'https://github.com/ascopes/jct.git/': The requested URL returned error: 403_

I really need help understanding all of this "rebase" and "stash" things. I also wanted to say that I aprecciate that you are taking the time to help me understand how Pull Requests and all of this work.

adolfo-trocoli commented:
Ok, no problem. I am very sorry I am being so slow but as I told you this has been my finals week. I finish my exams tomorrow and then I'll fix this and push it and will continue working on small issues on your project if that is ok with you.

ascopes commented:
Assume you no longer want to work on this?
